src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -187,123 +187,130 @@ state switch
                             {
                                 // Scenario 2.
-                                compilationPair = compilationPair.AddSyntaxTree(tree);
+                                pendingActions = [(inProgressProject, new CompilationAndGeneratorDriverTranslationAction.AddDocumentsAction([docState]))];
                                 inProgressProject = inProgressProject.AddDocuments([docState]);
                             }
                             else
                             {
-                                // Scenario 3.
-                                compilationPair = compilationPair.ReplaceSyntaxTree(oldTree, tree);
-
-                                // The old tree came from some other document with a different ID then we started with -- if the document ID still existed we would have
-                                // been in the Scenario 1 case instead. We'll find the old document ID, remove that state, and then add ours.
-                                var oldDocumentId = DocumentState.GetDocumentIdForTree(oldTree);
-                                Contract.ThrowIfNull(oldDocumentId, $"{nameof(oldTree)} came from the compilation produced by the workspace, so the document ID should have existed.");
-                                inProgressProject = inProgressProject
-                                    .RemoveDocuments([oldDocumentId])
+                                // Scenario 3. The old doc came from some other document with a different ID then we started
+                                // with -- if the document ID still existed we would have been in the Scenario 1 case
+                                // instead. We'll find the old document ID, remove that state, and then add ours.
+                                var inProgressWithDocumentRemoved = inProgressProject.RemoveDocuments([oldState.Id]);
+                                pendingActions = [
+                                    (inProgressProject, new CompilationAndGeneratorDriverTranslationAction.RemoveDocumentsAction([oldState])),
+                                (inProgressWithDocumentRemoved, new CompilationAndGeneratorDriverTranslationAction.AddDocumentsAction([docState]))];
+                                inProgressProject = inProgressWithDocumentRemoved
                                     .AddDocuments([docState]);
                             }
                         }
                     }
 
-                    // At this point, we now absolutely should have our tree in the compilation
-                    Contract.ThrowIfFalse(compilationPair.CompilationWithoutGeneratedDocuments.ContainsSyntaxTree(tree));
-                }
+                    // Transition us to a frozen in progress state.  With the best compilations up to this point, and the
+                    // pending actions we'll need to perform on it to get to the final state.
+                    var inProgressState = InProgressState.Create(
```

Comment:
this is core part of the change.  Instead of going to be in a final state, where we manually did all the steps ourselves, we go back to a normal frozen in-progress-state where there are still 1-2 pending translations to perform in the standard fashion.